File: examples/java/src/main/java/org/catholicos/QueryRunner.java
package org.catholicos;

import org.apache.jena.query.*;
import org.apache.jena.rdf.model.Model;
import org.apache.jena.rdf.model.ModelFactory;
import org.apache.jena.riot.RDFDataMgr;
import org.apache.jena.riot.RDFFormat;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Catholic Semantic Canon - SPARQL Query Runner
 *
 * This application loads the ontology and runs SPARQL queries from the queries/ folder.
 *
 * Usage:
 *     mvn exec:java                                    # Interactive menu
 *     mvn exec:java -Dexec.args="01"                   # Run query 01-list-all-classes.rq
 *     mvn exec:java -Dexec.args="--list"              # List available queries
 *     mvn exec:java -Dexec.args="--query 'SELECT...'" # Run inline SPARQL
 */
public class QueryRunner {

    private static final Path PROJECT_ROOT = findProjectRoot();
    private static final Path SOURCES_DIR = PROJECT_ROOT.resolve("sources");
    private static final Path QUERIES_DIR = PROJECT_ROOT.resolve("queries");
    private static final Path DEFAULT_ONTOLOGY = SOURCES_DIR.resolve("ontology-semantic-canon.ttl");

    private Model model;

    public static void main(String[] args) {
        QueryRunner runner = new QueryRunner();
        runner.run(args);
    }

    private static Path findProjectRoot() {
        // Start from current directory and look for sources/ folder
        Path current = Paths.get("").toAbsolutePath();

        // If we're in examples/java, go up two levels
        if (current.endsWith("java") || current.toString().contains("examples")) {
            current = current.getParent();
            if (current != null && current.endsWith("examples")) {
                current = current.getParent();
            }
        }

        // Check if sources exists here
        if (Files.exists(current.resolve("sources"))) {
            return current;
        }

        // Try going up from current directory
        while (current != null) {
            if (Files.exists(current.resolve("sources"))) {
                return current;
            }
            current = current.getParent();
        }

        // Fallback to current directory
        return Paths.get("").toAbsolutePath();
    }

    public void run(String[] args) {
        List<String> argList = Arrays.asList(args);

        // Handle --list flag
        if (argList.contains("--list") || argList.contains("-l")) {
            listQueries();
            return;
        }

        // Handle --query flag
        int queryIdx = argList.indexOf("--query");
        if (queryIdx == -1) queryIdx = argList.indexOf("-q");
        if (queryIdx != -1 && queryIdx + 1 < argList.size()) {
            loadOntology(DEFAULT_ONTOLOGY);
            runInlineQuery(argList.get(queryIdx + 1));
            return;
        }

        // Load ontology for all other operations
        loadOntology(DEFAULT_ONTOLOGY);

        // Run specific query by number
        if (!argList.isEmpty() && !argList.get(0).startsWith("-")) {
            String queryNumber = argList.get(0);
            runQueryByNumber(queryNumber);
            return;
        }

        // Interactive mode
        interactiveMenu();
    }

    private void loadOntology(Path ontologyPath) {
        System.out.println("Loading ontology from " + ontologyPath.getFileName() + "...");

        model = ModelFactory.createDefaultModel();

        try {
            RDFDataMgr.read(model, ontologyPath.toString());
            System.out.println("Loaded " + model.size() + " triples.");
        } catch (Exception e) {
            System.err.println("Error loading ontology: " + e.getMessage());
            System.err.println("Expected path: " + ontologyPath);
            System.err.println("Make sure you're running from the project directory.");
            System.exit(1);
        }
    }

    private List<Path> getQueryFiles() {
        if (!Files.exists(QUERIES_DIR)) {
            System.err.println("Queries directory not found: " + QUERIES_DIR);
            return Collections.emptyList();
        }

        try {
            return Files.list(QUERIES_DIR)
                    .filter(p -> p.toString().endsWith(".rq"))
                    .sorted()
                    .collect(Collectors.toList());
        } catch (IOException e) {
            System.err.println("Error listing queries: " + e.getMessage());
            return Collections.emptyList();
        }
    }

    private void listQueries() {
        System.out.println("\nAvailable queries:");
        for (Path query : getQueryFiles()) {
            System.out.println("  " + query.getFileName());
        }
    }

    private void runQueryByNumber(String number) {
        List<Path> queries = getQueryFiles();
        Optional<Path> matching = queries.stream()
                .filter(q -> q.getFileName().toString().startsWith(number))
                .findFirst();

        if (matching.isPresent()) {
            runQueryFile(matching.get());
        } else {
            System.err.println("No query found matching '" + number + "'");
            System.err.println("Use --list to see available queries");
        }
    }

    private void runQueryFile(Path queryPath) {
        System.out.println("\n=== Running: " + queryPath.getFileName() + " ===\n");

        try {
            String queryText = Files.readString(queryPath);
            executeQuery(queryText);
        } catch (IOException e) {
            System.err.println("Error reading query file: " + e.getMessage());
        }
    }

    private void runInlineQuery(String queryText) {
        System.out.println("\n=== Running inline query ===\n");
        executeQuery(queryText);
    }

    private void executeQuery(String queryText) {
        try {
            Query query = QueryFactory.create(queryText);

            if (query.isConstructType()) {
                // CONSTRUCT query
                try (QueryExecution qe = QueryExecutionFactory.create(query, model)) {
                    Model resultModel = qe.execConstruct();
                    System.out.println("--- CONSTRUCT Results (Turtle format) ---\n");
                    RDFDataMgr.write(System.out, resultModel, RDFFormat.TURTLE_PRETTY);
                }
            } else if (query.isSelectType()) {
                // SELECT query
                try (QueryExecution qe = QueryExecutionFactory.create(query, model)) {
                    ResultSet results = qe.execSelect();

                    // Print header
                    List<String> vars = results.getResultVars();
                    System.out.println(String.join(" | ", vars));
                    System.out.println("-".repeat(vars.size() * 20));

                    int count = 0;
                    while (results.hasNext()) {
                        QuerySolution sol = results.next();
                        List<String> values = new ArrayList<>();
                        for (String var : vars) {
                            if (sol.get(var) != null) {
                                String val = sol.get(var).toString();
                                // Shorten URIs for readability
                                if (val.startsWith("http")) {
                                    int lastSlash = val.lastIndexOf('/');
                                    int lastHash = val.lastIndexOf('#');
                                    int idx = Math.max(lastSlash, lastHash);
                                    if (idx > 0) val = val.substring(idx + 1);
                                }
                                values.add(val);
                            } else {
                                values.add("");
                            }
                        }
                        System.out.println(String.join(" | ", values));
                        count++;
                    }
                    System.out.println("\n(" + count + " results)");
                }
            } else if (query.isAskType()) {
                // ASK query
                try (QueryExecution qe = QueryExecutionFactory.create(query, model)) {
                    boolean result = qe.execAsk();
                    System.out.println("Result: " + result);
                }
            }
        } catch (Exception e) {
            System.err.println("Query error: " + e.getMessage());
        }
    }

    private void interactiveMenu() {
        List<Path> queries = getQueryFiles();
        if (queries.isEmpty()) return;

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\n" + "=".repeat(50));
            System.out.println("Available SPARQL Queries:");
            System.out.println("=".repeat(50));

            for (int i = 0; i < queries.size(); i++) {
                String name = queries.get(i).getFileName().toString();
                name = name.substring(0, name.length() - 3); // Remove .rq
                System.out.printf("  %2d. %s%n", i + 1, name);
            }

            System.out.println("\n  q. Quit");
            System.out.println("  c. Custom SPARQL query");
            System.out.println("=".repeat(50));

            System.out.print("\nSelect a query number (or 'q' to quit): ");
            String choice = scanner.nextLine().trim().toLowerCase();

            if (choice.equals("q")) {
                System.out.println("Goodbye!");
                break;
            } else if (choice.equals("c")) {
                System.out.println("\nEnter your SPARQL query (end with an empty line):");
                StringBuilder queryBuilder = new StringBuilder();
                String line;
                while (!(line = scanner.nextLine()).isEmpty()) {
                    queryBuilder.append(line).append("\n");
                }
                if (queryBuilder.length() > 0) {
                    executeQuery(queryBuilder.toString());
                }
            } else {
                try {
                    int idx = Integer.parseInt(choice) - 1;
                    if (idx >= 0 && idx < queries.size()) {
                        runQueryFile(queries.get(idx));
                    } else {
                        System.out.println("Invalid selection.");
                    }
                } catch (NumberFormatException e) {
                    System.out.println("Please enter a number.");
                }
            }
        }
    }
}

File: examples/python/query_runner.py
#!/usr/bin/env python3
"""
Catholic Semantic Canon - SPARQL Query Runner

This script loads the ontology and runs SPARQL queries from the queries/ folder.

Usage:
    python query_runner.py                     # Interactive menu
    python query_runner.py 01                  # Run query 01-list-all-classes.rq
    python query_runner.py --list              # List available queries
    python query_runner.py --query "SELECT ..."  # Run inline SPARQL
    python query_runner.py --semantic          # Use semantic queries (property paths)
    python query_runner.py --semantic 01       # Run semantic query 01
    python query_runner.py --semantic --list   # List semantic queries
"""

import argparse
import pickle
import sys
from pathlib import Path

try:
    from rdflib import Graph
except ImportError:
    print("Error: rdflib is not installed.")
    print("Install it with: pip install -r requirements.txt")
    sys.exit(1)


# Paths relative to this script
SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent.parent
SOURCES_DIR = PROJECT_ROOT / "sources"
QUERIES_DIR = PROJECT_ROOT / "queries"
RDFLIB_QUERIES_DIR = QUERIES_DIR / "rdflib"

# Default ontology file (OWL/XML format with pickle caching for stability)
DEFAULT_ONTOLOGY = SOURCES_DIR / "ontology-semantic-canon.owl"


def load_ontology(ontology_path: Path = DEFAULT_ONTOLOGY) -> Graph:
    """Load the ontology into an RDF graph, using pickle cache if available."""
    cache_path = ontology_path.with_suffix(".pickle")

    # Try to load from cache first
    if cache_path.exists():
        try:
            print(f"Loading from cache {cache_path.name}...")
            with open(cache_path, "rb") as f:
                g = pickle.load(f)
            print(f"Loaded {len(g)} triples from cache.")
            return g
        except Exception as e:
            print(f"Cache load failed ({e}), parsing ontology...")

    print(f"Loading ontology from {ontology_path.name}...")
    g = Graph()

    # Determine format from extension
    suffix = ontology_path.suffix.lower()
    format_map = {
        ".ttl": "turtle",
        ".owl": "xml",
        ".owx": "xml",
        ".rdf": "xml",
        ".n3": "n3",
        ".nt": "nt",
    }
    rdf_format = format_map.get(suffix, "turtle")

    g.parse(str(ontology_path), format=rdf_format)
    print(f"Loaded {len(g)} triples.")

    # Save to cache for faster subsequent loads
    try:
        print(f"Saving to cache {cache_path.name}...")
        with open(cache_path, "wb") as f:
            pickle.dump(g, f)
        print("Cache saved successfully.")
    except Exception as e:
        print(f"Warning: Could not save cache: {e}")

    return g


def list_queries(semantic: bool = False) -> list[Path]:
    """List all available query files.

    Args:
        semantic: If True, list queries from the rdflib/ subdirectory
                  which use semantic patterns optimized for rdflib/Python.
    """
    query_dir = RDFLIB_QUERIES_DIR if semantic else QUERIES_DIR

    if not query_dir.exists():
        print(f"Error: Queries directory not found: {query_dir}")
        return []

    queries = sorted(query_dir.glob("*.rq"))
    return queries


def run_query(graph: Graph, query_text: str) -> None:
    """Execute a SPARQL query and print results."""
    try:
        results = graph.query(query_text)

        # Check if it's a CONSTRUCT query (returns a graph)
        if hasattr(results, 'graph') and results.graph is not None:
            print("\n--- CONSTRUCT Results (Turtle format) ---\n")
            print(results.graph.serialize(format="turtle"))
        else:
            # SELECT query results
            # Defensive check for results.vars (rdflib bug workaround)
            if results.vars and hasattr(results.vars, '__iter__'):
                try:
                    # Print header
                    header = " | ".join(str(var) for var in results.vars)
                    print(f"\n{header}")
                    print("-" * len(header))
                except TypeError:
                    # results.vars is not iterable due to rdflib bug
                    print("\n(Unable to display header due to rdflib bug)")

            count = 0
            for row in results:
                values = []
                for val in row:
                    if val is None:
                        values.append("")
                    else:
                        # Shorten URIs for readability
                        val_str = str(val)
                        if isinstance(val_str, str) and val_str.startswith("http"):
                            val_str = val_str.split("/")[-1].split("#")[-1]
                        values.append(str(val_str))
                print(" | ".join(values))
                count += 1

            print(f"\n({count} results)")

    except Exception as e:
        print(f"Query error: {e}")
        print("(This may be an intermittent rdflib/Python 3.12 bug - try running the query again)")


def run_query_file(graph: Graph, query_path: Path) -> None:
    """Load and execute a query from a file."""
    print(f"\n=== Running: {query_path.name} ===\n")

    with open(query_path, "r") as f:
        query_text = f.read()

    run_query(graph, query_text)


def interactive_menu(graph: Graph, semantic: bool = False) -> None:
    """Show an interactive menu to select and run queries.

    Args:
        semantic: If True, show semantic queries from queries/rdflib/
                  instead of regex-based queries from queries/.
    """
    queries = list_queries(semantic=semantic)

    if not queries:
        return

    query_type = "Semantic" if semantic else "Standard"

    while True:
        print("\n" + "=" * 60)
        print(f"Available {query_type} SPARQL Queries:")
        if semantic:
            print("(Optimized for rdflib - use queries/jena/ with arq for full features)")
        print("=" * 60)

        for i, q in enumerate(queries, 1):
            # Extract description from filename
            name = q.stem
            print(f"  {i:2}. {name}")

        print("\n  q. Quit")
        print("  c. Custom SPARQL query")
        print(f"  s. Switch to {'standard' if semantic else 'semantic'} queries")
        print("=" * 60)

        choice = input("\nSelect a query number (or 'q' to quit): ").strip().lower()

        if choice == 'q':
            print("Goodbye!")
            break
        elif choice == 'c':
            print("\nEnter your SPARQL query (end with an empty line):")
            lines = []
            while True:
                line = input()
                if line == "":
                    break
                lines.append(line)
            if lines:
                run_query(graph, "\n".join(lines))
        elif choice == 's':
            # Switch query type and refresh the menu
            semantic = not semantic
            queries = list_queries(semantic=semantic)
            query_type = "Semantic" if semantic else "Standard"
            print(f"\nSwitched to {query_type.lower()} queries.")
        else:
            try:
                idx = int(choice) - 1
                if 0 <= idx < len(queries):
                    run_query_file(graph, queries[idx])
                else:
                    print("Invalid selection.")
            except ValueError:
                print("Please enter a number.")


def main():
    parser = argparse.ArgumentParser(
        description="Run SPARQL queries against the Catholic Semantic Canon ontology"
    )
    parser.add_argument(
        "query_number",
        nargs="?",
        help="Query number to run (e.g., '01' for 01-list-all-classes.rq)"
    )
    parser.add_argument(
        "--list", "-l",
        action="store_true",
        help="List available queries"
    )
    parser.add_argument(
        "--query", "-q",
        type=str,
        help="Run an inline SPARQL query"
    )
    parser.add_argument(
        "--ontology", "-o",
        type=Path,
        default=DEFAULT_ONTOLOGY,
        help=f"Path to ontology file (default: {DEFAULT_ONTOLOGY.name})"
    )
    parser.add_argument(
        "--semantic", "-s",
        action="store_true",
        help="Use semantic queries (UNION patterns to emulate subClassOf*, avoids rdflib segfaults)"
    )

    args = parser.parse_args()

    # List queries mode
    if args.list:
        queries = list_queries(semantic=args.semantic)
        query_type = "semantic" if args.semantic else "standard"
        print(f"\nAvailable {query_type} queries:")
        for q in queries:
            print(f"  {q.name}")
        if not args.semantic:
            print("\n  (Use --semantic to list semantic queries)")
        return

    # Load ontology
    if not args.ontology.exists():
        print(f"Error: Ontology file not found: {args.ontology}")
        print("Make sure you're running from the examples/python directory")
        print("or provide the path with --ontology")
        sys.exit(1)

    graph = load_ontology(args.ontology)

    # Run inline query
    if args.query:
        run_query(graph, args.query)
        return

    # Run specific query by number
    if args.query_number:
        queries = list_queries(semantic=args.semantic)
        matching = [q for q in queries if q.name.startswith(args.query_number)]
        if matching:
            run_query_file(graph, matching[0])
        else:
            query_type = "semantic" if args.semantic else "standard"
            print(f"No {query_type} query found matching '{args.query_number}'")
            print("Use --list to see available queries")
        return

    # Interactive mode
    interactive_menu(graph, semantic=args.semantic)


if __name__ == "__main__":
    main()
